### PR TITLE
Add functionality for gating feature flags on epochs ; rejigger epoch lints

### LINKED ISSUE #48794: Make it possible to gate features on epochs
Features like NLL and dyn trait should automatically turn on in epoch 2018.

### PR DESCRIPTION
**Pull request by Manishearth**
fixes #48794

r? @nikomatsakis

### CODE REVIEW

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
Looks nice!

**Comment by Manishearth** [comment]
Added test.

Oh, also, I'm going to make bare_trait_object and unreachable_pub not be epoch lints -- I'll keep epoch lints to be actual breaking changes like tyvar_rptr and the keywords, and these lints can go ungrouped for now (we can group them later if we wish?)

File: src/test/run-pass/epoch-gate-feature.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+// Checks if the correct registers are being used to pass arguments
+// when the sysv64 ABI is specified.
+
+// compile-flags: -Zepoch=2018
+
+pub trait Foo {}
+
+// should compile without the dyn trait feature flag
+fn foo(x: &dyn Foo) {}
```
Comment (nikomatsakis) [inline src/test/run-pass/epoch-gate-feature.rs]: <3

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**